### ISSUE TITLE
Incremental pruning reorders module hooks

### ISSUE THREAD

**Issue by sharvil** [pytorch/pytorch#37138]
## To Reproduce

1. Add identity pruning to a module's weight to reparameterize it
1. Call `register_forward_pre_hook` to register additional hooks to run on the module after reparameterization (e.g. additional reparameterizations)
1. Add a pruning schedule to the module's weight (e.g. prune 10% weights every epoch)
1. Notice that pruning reorders hooks

## Expected behavior

Once a module's weight has been reparameterized for pruning (either through identity pruning or otherwise), additional weight pruning applied to the same (module, weight) pair should not result in that module's hooks being reordered. Preserving hook order allows incremental pruning to compose with other features, e.g. weight normalization.

**Comment by sharvil** [comment]
@mickypaganini, could you please take a look? Thanks.

**Comment by zhangguanheng66** [comment]
could you share a simple code snippet to reproduce the error?

**Comment by sharvil** [comment]
Adapted from the unit test:

```python
import torch
import torch.nn as nn
import torch.nn.utils.prune as prune

def test_muliple_pruning_calls_preserve_hooks():
    # Initial setup: a module with 3 pre-forward hooks where the middle one is
    # the hook for pruning.
    m = nn.Conv3d(2, 2, 2)
    m.register_forward_pre_hook(lambda x, y: y)
    prune.l1_unstructured(m, name='weight', amount=0.1)
    m.register_forward_pre_hook(lambda x, y: y)

    original_hooks = list(m._forward_pre_hooks.keys())
    prune.l1_unstructured(m, name='weight', amount=0.1)
    new_hooks = list(m._forward_pre_hooks.keys())

    print(original_hooks)
    print(new_hooks)

test_muliple_pruning_calls_preserve_hooks()
```

Expected output:
`[0, 1, 2]`

Actual output:
`[0, 2, 3]`

**Comment by mickypaganini** [comment]
If I understand correctly, you want to apply three reparametrizations of different kinds: A, B, C.
Say A and C are pruning reparametrizations, and B is a reparametrization of a different kind.

What we currently do, which I agree may not be safe or correct, is to fold A+C=D, and then apply: B, D.

What you are suggesting (correct me if I'm wrong), is equally unsafe and potentially incorrect, is to fold A+C=D, and then apply: D, B. 

Neither is ok. We should apply A, then B, then C in the correct order because we cannot assume, in general, that these operations commute.

However, the even bigger issue here is that we cannot assume that different reparametrizations will even be compatible with each other. In other words, B may not be applicable to a pruned module, and, vice versa, a module reparametrized by B may not be prunable.

How is `parametrize` handling this situation @Lezcano @apaszke @albanD ?

**Comment by Lezcano** [comment]
The `Parametrization` framework will handle this issue correctly. Two `Pruning` methods will only be folded into one if they were applied on the same weight one after the other with no other `Parametrization`s between them.

This particular problem with the hooks will not be present, as `Parametrization`s (`Pruning` included) will not make use of hooks.

**Comment by sharvil** [comment]
@mickypaganini I agree that these operations don't necessarily commute and that neither of the folding options is ideal.

Let me explain where I'm coming from. I have a pruning schedule that drops a larger percentage of units over (training) time. I also have another reprameterization I perform after pruning. I'd like to preserve the sequence of operations I've defined on the tensor in question (prune -> reparameterize -> forward) while still being able to increase the number of pruned units. The pruning fold operation in a way allows me to achieve that – as long as it doesn't reorder the pre-forward hooks. That's the scope of this PR.

@Lezcano, will the `Parameterization` framework allow me to insert a new `Parameterization` at an arbitrary location? It would be really nice to have more control over whether a pruning method gets folded or not, and which location it occupies in the reparameterization stack.

**Comment by Lezcano** [comment]
It will indeed.

**Comment by mickypaganini** [comment]
@Lezcano That's not the problem. That, in the meantime, could be achieved with an easy fix with forward pre hooks as well, given that it's an ordered dictionary. We can easily make it such that two pruning methods are folded together only if they are applied back-to-back. That's fine. @sharvil we can support your use case with a quick PR.

The bigger issue I was wondering about it what to do when two reparametrizations of different kinds are applied that cannot be easily combined. That's a larger problem to address beyond @sharvil's specific issue and use case.

There isn't an algebra of reparametrizations that is well defined. Reparametrizations can't be fully aware of each other and we can't make sure that, in the general case, they are compatible.

As an example, consider weight norm, where we do:
```
[1] W |--> v = Param(W), g = Param(norm(W)), W = Attribute(g * v)
```
This reparametrization requires the creation of two parameters from a single original parameter. Fine on its own.

Pruning does something similar:
```
[2] W |--> W_orig = Param(W), M = Buffer(get_some_mask()), W = Attribute(W_orig * M)
```

Now, what if we want to apply weight norm to a pruned weight (something which is in theory totally valid)?

In that case, when weight norm tries to do `v = Param(W)`, weight norm will try to make a param out of something that is itself is a derived expression, `W_orig * M`. So, instead if the original in [1], we need to do:
```
W |--> v = Attribute(W), g = Attribute(norm(W)), W = Attribute(g * v),
```
which, when applied to a pruned weight, will be doing:
```
W |--> v = Attribute(W_orig * M), g = Attribute(norm(W_orig * M)), W = Attribute(g * v),
```
which will allow the gradient to be propagated correctly to `W_orig`.

This would mean that every reparametrization needs to be aware of the logic of all other reparametrizations.

**Comment by Lezcano** [comment]
A `Parametrization` just needs to be able to evaluate the chain of previously applied parametrizations. 

The current proposal for the parametrization framework allows to do these kind of things out of the box.

I am not completely sure what you mean by "low-rank factorization swap". When it comes to low-rank models, I was able to implement a Parametrization that makes a weight into a rank `k` matrix via a factorisation into two orthogonal matrices and the singular values in this framework, but this might not be what you are referring to...

**Comment by mickypaganini** [comment]
It doesn't just need to evaluate them. I think that applying a reparametrization to something that is itself a derived quantity is not always safe/trivial.

**Comment by Lezcano** [comment]
The current implementation allows to do so via a chaining mechanism, where parametrisations are chained to each other. `Parametrization`s may be merged as currently done in the  pruning implementation by overriding the `chain` method and implementing there the merge.

I am not aware of any case where the current design does not work, but I might have missed something. Do you know of any case where it fails?

**Comment by sharvil** [comment]
> In that case, when weight norm tries to do `v = Param(W)`, weight norm will try to make a param out of something that is itself is a derived expression, `W_orig * M`. So, instead if the original in [1], we need to do:
> 
> ```
> W |--> v = Attribute(W), g = Attribute(norm(W)), W = Attribute(g * v),
> ```
> 
> which, when applied to a pruned weight, will be doing:
> 
> ```
> W |--> v = Attribute(W_orig * M), g = Attribute(norm(W_orig * M)), W = Attribute(g * v),
> ```
> 
> which will allow the gradient to be propagated correctly to `W_orig`.
> 
> This would mean that every reparametrization needs to be aware of the logic of all other reparametrizations.

@mickypaganini, in this example, could weight norm look at the type of the weight it's reparameterizing to decide whether it should create new `Parameter`s? If the weight is a `Tensor`, it creates new computed attributes and if it's a `Parameter`, it has to create new `Parameter`s since overwriting the weight property will replace an existing `Parameter`. That logic doesn't depend on any knowledge of upstream reparameterizations.

**Comment by mickypaganini** [comment]
Correct. It would have to be modified to do the right thing, depending on the state the module has been put into by previous reparametrizations -- in that sense, it has to have knowledge of them. Then, yes, in practice, at least in this case, this would simply mean creating different logical pathways in weight norm (and in pruning too, for that matter) to allow the reparametrization of a parameters, or an attribute, or a buffer, or whatever else the tensor can be. Not elegant, but doable.

So, suggested changes for now are:
- modifying `_get_composite_method` to only fold pruning methods into a container if they are applied back-to-back (change the loop through hooks to a check on the last hook).
   - this will now imply the coexistence of as many separate masks per module as there are pruning hooks, and the need to correctly map from each pruning hook to its mask.
   - we might have to think more carefully about how removing reparametrizations would work, and whether in can only be done in reverse order of application to avoid "holes" in the reparametrization history...
- modifying weight norm and pruning (are there any other reparametrizations of this sort in pytorch at the moment?) to be able to reparametrize a tensor that is not a parameter.
- extensive testing to ensure correctness and compatibility with models like RNNs.

Before implementing all these changes, @apaszke @soumith do you have any thoughts on this?

**Comment by sharvil** [comment]
@mickypaganini, do your proposed changes handle the situation I described above? Given a sequence of operations (prune -> reparameterize -> forward), I would like to increase the pruning rate as a function of training step by applying additional pruning methods between (prune -> reparameterize).

It seems to me that automatic folding and `remove` are incompatible with each other. If I apply pruning in the order (prune -> reparameterize -> prune), I can remove the pruning methods independently. If I apply (prune -> prune -> reparameterize), I silently lose that ability.

In my ideal world, each reparameterization would be a subclass of some `BaseReparameterization` class that has a `remove` method. The subclass provides explicit operations for the type of reparameterization in question (e.g. `fold(method : PruningMethod)`). A user gets an instance of the subclass when they apply a reparameterization and they can choose to hold on to that reference if they'd like to perform operations on that reparameterization in the future, or not.

**Comment by sharvil** [comment]
Ping. Any update on this issue? Can we move forward with or reject the PR?

**Comment by albanD** [comment]
Hi,

Is this still going to be relevant after the refactor of pruning within the reparametrization module?

**Comment by Lezcano** [comment]
This problem will not be present when we implement the `Pruning` methods as a subclass of `Parametrization`.

**Comment by sharvil** [comment]
Glad to to hear it. I'm going to drop this PR and bug. I've implemented my own pruning module to work around this issue in the meantime so I'm unblocked.
